clientconn.go

```diff
@@ -1020,17 +1040,30 @@ func (ac *addrConn) createTransport(connectRetryNum, ridx int, backoffDeadline,
 		close(ac.ready)
 		ac.ready = nil
 	}
+	resetBackoff := ac.resetBackoff
 	ac.mu.Unlock()
 	timer := time.NewTimer(backoffDeadline.Sub(time.Now()))
 	select {
 	case <-timer.C:
+	case <-resetBackoff:
+		timer.Stop()
+		ac.mu.Lock()
+		ac.connectRetryNum = 0
+		ac.mu.Unlock()
 	case <-ac.ctx.Done():
 		timer.Stop()
 		return false, ac.ctx.Err()
 	}
 	return false, nil
 }
 
+func (ac *addrConn) resetConnectBackoff() {
+	ac.mu.Lock()
```

Comment:
Also set `ac.connectRetryNum = 0`?

Reply:
Yes, I could do that here instead of line 1051 above, which would mean we will reset the backoff deadlines even if we aren't currently backing off.  Done.